### PR TITLE
[DI] Remove skipping magic for autowired methods

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no (master only)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Wildcard based autowiring made it required to auto-skip methods that were not wireable.
Now that things need to be explicit (ie via the `@required` annotation, or via configuration), this "automagic" behavior is not required anymore.
Since it can lead to wtf moments ("*I* did *put that `@required` annotation, why is it ignored by autowiring?*"), I think we should remove it.
This also fixes another issue where configured method calls had their optional arguments wired, while we want only the constructor's to behave as such.

### CODE REVIEW

**Comment by weaverryan** [comment]
For reference, here are the original decisions we made about how to try to autowire setter methods, but not fail as often / consistently when compared with constructor autowiring - items 2, 3 and 4 in my comment: https://github.com/symfony/symfony/pull/17608#issuecomment-180837991

I think this PR is wonderful! The `@required` completely removes the need for us to try to make this decision for the user.

@nicolas-grekas the only thing I noticed when trying this was that if I had `@required` above a method with *zero* arguments, that method wasn't called. With this philosophy, it now *feels* like this method should also be called. However, if I have a method with a single optional argument that cannot be autowired, it *feels* like that one should *not* be called (which is how it works now). But those are kind of the same situation (i.e. should we call a method when we have nothing to pass to it), so I'm not totally sure :).

Thanks!

**Comment by fabpot** [comment]
@weaverryan As `@required` is added by the user, he knows what he wants. So to me, the methods marked with `@required` must always be called, regardless of their arguments.

**Comment by weaverryan** [comment]
Nicolas and I were talking off-GitHub, and we agree - we should always call methods with `@required`, unless they cannot be called due to non-autowireable args (then we should throw an exception). 

What I forgot was that a method with *no* arguments is *getter*-injected. In other words, it's incorrect for me to assume that I could add `@required` above a method with no args (e.g. `configure()`) and expect it to be called. This is a little magic - that we're deciding between setter and getter injection based on the presence/absence of at least one arg -, but it seems like it needs to be done that way (and Nicolas and I discussed how we should perhaps throw an exception if there is an `@required` method with no args, but also not return type - so at least the user will get a clear error).

**Comment by fabpot** [comment]
I think throwing an exception when there is a `@required` annotation without arguments is a good idea indeed.

**Comment by nicolas-grekas** [comment]
Exceptions added, all code related to skipping logic removed.
I also reverted #21602: abstract getter don't need to be autowired anymore, we'd better encourage using the annotation instead.

File: src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php [hunk 1]
```diff
@@ -246,20 +233,24 @@ private function autowireCalls(\ReflectionClass $reflectionClass, array $methodC
...
      * @throws RuntimeException
      */
-    private function autowireMethod(\ReflectionMethod $reflectionMethod, array $arguments, $mode)
+    private function autowireMethod(\ReflectionMethod $reflectionMethod, array $arguments)
     {
-        $didAutowire = false; // Whether any arguments have been autowired or not
+        $isConstructor = $reflectionMethod->isConstructor();
+
+        if (!$isConstructor && !$arguments && !$reflectionMethod->getNumberOfRequiredParameters()) {
+            throw new RuntimeException(sprintf('Cannot autowire service "%s": method %s::%s() has only optional arguments, thus must be wired explicitly.', $this->currentId, $reflectionMethod->class, $reflectionMethod->name));
```
Comment (fabpot) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: This can also be hit when there is no args at all, right?

Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php]: a method with no argument at all is considered as a getter, thus can't hit this method, see
https://github.com/symfony/symfony/pull/22030/files#diff-62df969ae028c559d33ffd256de1ac49R219

**Pull request merged**